### PR TITLE
apis for application gateway

### PR DESCRIPTION
**Pull request by kmarquardsen**

### CODE REVIEW

File: pkg/app_gtwy/server/aether-app-gtwy-impl.go [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+			ip := string(imsiMap[imsi].Metric["ip"])
+			newDev.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			newDev.Attached = &attached
+		}
+
+		listDeviceResponse = append(listDeviceResponse, newDev)
+	}
+
+	log.Infof("GetAppGtwyDevices")
```
Comment (SeanCondon) [inline pkg/app_gtwy/server/aether-app-gtwy-impl.go:81]: I think it's always good to include some dynamic info in a log message - e.g. even the number of devices would be a help

File: pkg/app_gtwy/server/aether-app-gtwy-impl.go [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+		}
+
+		imsi := strconv.Itoa(int(*s.Imsi))
+		if _, ok := imsiMap[imsi]; ok {
+			ip := string(imsiMap[imsi].Metric["ip"])
+			device.Ip = &ip
+			attached := imsiMap[imsi].Value.String()
+			device.Attached = &attached
+		}
+	}
```
Comment (SeanCondon) [inline pkg/app_gtwy/server/aether-app-gtwy-impl.go:128]: if you did find the device, you could return here (to save looping around again)

File: pkg/app_gtwy/server/aether-app-gtwy-impl.go [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+
+	log.Infof("GetDevice")
+	return ctx.JSON(http.StatusOK, device)
```
Comment (SeanCondon) [inline pkg/app_gtwy/server/aether-app-gtwy-impl.go:131]: I think if you return inside the loop, then you can conclude that if you get to here the device was not found and you can return an error

File: pkg/app_gtwy/server/aether-app-gtwy-impl.go [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+
+	// Property: device-group []EnterprisesEnterpriseSiteDeviceGroup
+	// Handle []Object
+	deviceGroups := make([]externalRef0.EnterprisesEnterpriseSiteDeviceGroup, 0)
+	reflectEnterprisesEnterpriseSiteDeviceGroup, err := utils.FindModelPluginObject(d.device, "EnterprisesEnterpriseSiteDeviceGroup", params...)
+	if err != nil {
+		return nil, err
+	}
+	if reflectEnterprisesEnterpriseSiteDeviceGroup != nil {
+		for _, key := range reflectEnterprisesEnterpriseSiteDeviceGroup.MapKeys() {
```
Comment (SeanCondon) [inline pkg/app_gtwy/server/aether-app-gtwy-impl.go:224]: nice - good to see you got the hang of this

File: pkg/app_gtwy/server/aether-app-gtwy-impl.go [hunk 5]
```diff
@@ -0,0 +1,684 @@
...
+		return nil, fmt.Errorf("error unmarshalling gnmiResponse %v", err)
+	}
+	mpd := ModelPluginDevice{
+		device: gnmiResponse,
+	}
+
+	return mpd.toEnterprisesEnterpriseSite(args...)
+}
+
+// ModelPluginDevice - a wrapper for the model plugin
```
Comment (SeanCondon) [inline pkg/app_gtwy/server/aether-app-gtwy-impl.go:167]: you could split this Model Plugin related stuff out in to a separate file to keep the concepts cleaner

**Review by SeanCondon — APPROVED** [review]
I left a few minor comments, but this is great - approved - merge at your leisure

File: cmd/aether-roc-api/aether-roc-api.go [hunk 6]
```diff
@@ -57,6 +57,7 @@ func main() {
 	certPath := flag.String("certPath", "", "path to client certificate")
 	gnmiEndpoint := flag.String("gnmiEndpoint", "onos-config:5150", "address of onos-config")
 	gnmiTimeout := flag.Duration("gnmiTimeout", 10*time.Second, "timeout for the gnmi requests")
+	analyticsEndpoint := flag.String("analyticsEndpoint", "http://10.244.0.31:9090", "prometheus address")
```
[the remark this answers is not in the document]
Comment (kmarquardsen) [reply cmd/aether-roc-api/aether-roc-api.go:60]: Good eye, thanks

File: pkg/app_gtwy/server/aether-app-gtwy-impl.go (hunk 5 above)
Comment (kmarquardsen) [reply pkg/app_gtwy/server/aether-app-gtwy-impl.go:167]: Agreed!

**Pull request merged**